File: src/main/java/editor/database/attributes/OptionalAttribute.java
package editor.database.attributes;

/**
 * This class represents the value of a card attribute that might or might not
 * actually appear on a card, such as power, toughness, and loyalty. Mana cost
 * does not count, as an empty mana cost is still a valid mana cost.
 * 
 * @author Alec Roelke
 */
public interface OptionalAttribute
{
    /**
     * @return An unspecified attribute that is missing from a card.
     */
    public static OptionalAttribute empty()
    {
        // This isn't actually a functional interface, but this shorthand is more
        // convenient than fully defining an anonymous class
        return () -> false;
    }

    /**
     * @return <code>true</code> if the attribute is present, and <code>false</code>
     * otherwise.
     */
    boolean exists();    
}

File: src/main/scala/editor/database/attributes/OptionalAttribute.scala
package editor.database.attributes

/**
 * Global data about optional attributes.
 * @author Alec Roelke
 */
object OptionalAttribute {
  /** @return an [[OptionalAttribute]] with no value that's always missing. */
  def empty = new OptionalAttribute { override val exists = false }
}

/**
 * A card attribute that may or may not appear on a card, such as power or toughness which only appears on creatures
 * and Vehicles.  Note that [[ManaCost]] is not optional, as an empty mana cost is still a mana cost.
 * 
 * @author Alec Roelke
 */
trait OptionalAttribute {
  /** True if the attribute is defined on a card, and false otherwise. */
  def exists: Boolean
}